### DISCUSSION TITLE
Axe-core in the test setup: to leave vs to keep

### DISCUSSION THREAD

**Discussion by Chiemezuo** [bokeh/bokeh#14057] [Polls]
This discussion references PR #14032 

## Some Useful Things to Note
* Bokeh data visualizations are mostly rendered as a `<canvas>` element in HTML.
* Axe-core treats a `<canvas>` element pretty much like a simple node, and has no idea how to perform any accessibility checks on it.

On running all (~112 or so) visualizations in the [bokeh docs gallery](https://docs.bokeh.org/en/latest/docs/gallery.html) both via the axe Devtools extension in the browser and the axe-core library, I made some interesting findings:
1. For visualizations without the toolbar and/or widgets (e.g [simple image scatter](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/image_url.html), there were no reported issues whatsoever with axe-core. All passed.
2. For visualizations with a toolbar but no additional widget (e.g [grid bounds example](https://docs.bokeh.org/en/latest/docs/examples/styling/plots/grid_bounds.html), the primary issue axe-core finds is that the Bokeh icon has [no discernible text](https://dequeuniversity.com/rules/axe/4.10/link-name?application=AxeChrome)
3. For visualizations with a widget (such as a slider, dropdown, multichoice, etc), axe-core does a good job of finding accessibility issues. This is because those widgets tend to be outside the `canvas`, and can be checked for accessibility violations. This category of visualizations is where most of the violations occurred.

## Common Issues axe-core reported
[X] aria-label attribute does not exist or is empty.
[X] Contrast ration issues.
[X] Ensure links have discernible text.
[X] Form element does not have an implicit (wrapped) <label>.
[X] Form element does not have an explicit <label>.
[X] ARIA attribute is not allowed: aria-selected=\"true\"
[X] Form element has explicit <label> that is hidden"

## The Next Step
The above issues come up somewhat frequently for plots with a toolbar and a widget, and there's a question of whether these are enough to justify leaving axe-core in the test setup. It's important to note that axe-core will be even more meaningful if future modifications of bokeh will see to more things being removed from the canvas element and rendered in other ways. This brings me to the question I wanted to ask:

**Comment by bryevdv** [comment]
I guess I don't really have the context to comment, fully. For example: what exactly is axe-core? 

I think you are asking whether we should have a runs a tool CI as a test, or whether we should run the tool manually once and generate a report, but not check in any test machinery. Without knowing more, my offhand thought is that that a false dichotomy. I think we should have test machinery that we *could* run in CI, if we wanted to, regardless of whether we *do*.

As I suggested in the PR, we would ideally have machinery that lets the report be generated similar to running other existing tests:
```bash
pytest tests/accessibility 

# or

node make test/accessibility
```
or something along those lines. 

Then any one can easily produce a new report any time they want. Whether we actually run something in CI or not, is a separate question.

**Reply by Chiemezuo** [reply]
Hi, Bryan!
I'm sorry I didn't give as much context, but I'll try to do that both here, and in a much more detailed GitHub issue.

[Axe-core](https://github.com/dequelabs/axe-core) is an accessibility engine for web UI testing. It's a tool that runs the HTML (for the most part) of a website, and then identifies accessibility errors and the pertinent/responsible nodes or elements for those errors. It does this via the axe API.

For the question itself, it wasn't really about CI. It was just a matter of leaving axe-core in the PR or not. It has its usefulness up until a point, and I asked the question because I wasn't sure whether or not we would ever get beyond that point. I'll illustrate even further.

The nature of bokeh is to render data visualizations as `<canvas>` elements. Axe-core itself can't do anything with `<canvas>`  elements, but there are certain bokeh data visualizations that have some content **outside** the canvas elements. With those, it can be very helpful.

However, there aren't so many unique problems (I only count about 7 for now) that would necessarily warrant using axe-core right now, but there could be in the future. It's not a certainty, but it's a possibility. 

Of course, it can still be removed now and iteratively added later on, but I'll write a detailed report and also attach it to this discussion so that the issues it can currently identify would be easy to reference.

**Comment by Chiemezuo** [comment]
I suspect it'll be better to write down my findings also in this GitHub discussion, so it could be a central location for axe-core related discussions. I ran ~114 Bokeh data visualizations from the examples section of the documentation against axe-core, and I will thoroughly write down the issues that came up.

## Types of plots with accessibility issues axe-core can't find.
Take for instance the following [image URL scatter](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/image_url.html):
![image](https://github.com/user-attachments/assets/5148a67e-9371-42e1-acc3-06d8febeb60a)

All of the contents of that image scatter are wrapped in the `<canvas>` element, which axe-core doesn't know what to do with. Running that through axe-core generates no errors. Essentially, all simple plots without toolbars will follow a similar pattern.

Here are some more error-free visualizations in the documentation examples:
* [Scatter markers](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/markers.html)
* [Basic bar charts](https://docs.bokeh.org/en/latest/docs/examples/basic/bars/basic.html)
* [Fixed axis](https://docs.bokeh.org/en/latest/docs/examples/basic/axes/fixed_axis.html)
* [Intervals](https://docs.bokeh.org/en/latest/docs/examples/basic/bars/intervals.html)

There are several more, but they follow the general pattern of not having anything beyond the plot itself and the axes.

## Plots with Toolbars that trigger axe-core issues
Generally, the problem with toolbars is the inclusion of the Bokeh logo as an anchor `<a>` tag without [discernible text](https://dequeuniversity.com/rules/axe/4.10/link-name?application=AxeChrome). It could be fixed in a number of ways and would generally not appear again.

Look at the following picture from the [colour scatter](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/color_scatter.html) for a better illustration:

![Screenshot 2024-09-17 200009](https://github.com/user-attachments/assets/a3bb9811-ca89-4405-9405-5c502394684a)

The Bokeh logo has been highlighted. Axe-core was able to find this because the toolbar is in its own `<div>` element. 

Places in the documentation where this issue was found: [Colour scatter](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/color_scatter.html), [Basic scatter](https://docs.bokeh.org/en/latest/docs/examples/basic/scatters/elements.html), [Lorenz](https://docs.bokeh.org/en/latest/docs/examples/basic/lines/lorenz.html), [Linear c_map](https://docs.bokeh.org/en/latest/docs/examples/basic/data/linear_cmap.html), [cmap colorbar](https://docs.bokeh.org/en/latest/docs/examples/basic/data/linear_cmap_colorbar.html), [transform markers](https://docs.bokeh.org/en/latest/docs/examples/basic/data/transform_markers.html), [transform jitters](https://docs.bokeh.org/en/latest/docs/examples/basic/data/transform_jitter.html), [log plots](https://docs.bokeh.org/en/latest/docs/examples/basic/axes/logplot.html), [twin axes](https://docs.bokeh.org/en/latest/docs/examples/basic/axes/twin_axes.html), [stacked area](https://docs.bokeh.org/en/latest/docs/examples/basic/areas/stacked_area.html), [annotations legend](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/legend.html), [whisker annotations](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/whisker.html), [annotations span](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/span.html), [annotations slope](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/slope.html), [annotations box](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/box_annotation.html).

There are several more of these issues, but there's no added benefit of including them here because they have the same source of error, and one fix would immediately reduce axe-core issues here to zero.

## Plots with external elements beyond a toolbar.
These are the occurrences that I feel make a strong case for keeping axe-core in the test setup.

Let's start with the [data table plot](https://docs.bokeh.org/en/latest/docs/examples/interaction/linking/data_table_plot.html):
![image](https://github.com/user-attachments/assets/5dbe9fe7-d803-48c0-b1aa-697519c5f083)

It's a plot that generates a table outside itself, and from the highlighted issue, we can see right away that axe-core detects contrast issues between the text and the background where the text falls in. 

And over here in the [color sliders](https://docs.bokeh.org/en/latest/docs/examples/interaction/js_callbacks/color_sliders.html), we have an issue with sliders not having [accessible names](https://dequeuniversity.com/rules/axe/4.10/aria-input-field-name?application=AxeChrome). 

![image](https://github.com/user-attachments/assets/2113ba2b-5349-4323-9bea-326769939493)

It's also in the [js_callbacks sliders](https://docs.bokeh.org/en/latest/docs/examples/interaction/js_callbacks/slider.html)

![image](https://github.com/user-attachments/assets/94c1f85f-04f2-401e-8e06-65ffc2b26e80)


Yet another case would be [js_on_event](https://docs.bokeh.org/en/latest/docs/examples/interaction/js_callbacks/js_on_event.html) with a color contrast issue with buttons
![image](https://github.com/user-attachments/assets/1c7f19f7-b881-4d09-92be-e7dd4fbf56df)

For the [multiselect widget](https://docs.bokeh.org/en/latest/docs/examples/interaction/widgets/multiselect.html), we have an issue with the [select element not having an accessible name](https://dequeuniversity.com/rules/axe/4.10/select-name?application=AxeChrome)
![image](https://github.com/user-attachments/assets/bdf63c25-b7c9-48d9-97c3-6a45fa193ff6)

For multichoice, an[ unsupported ARIA attribute](https://dequeuniversity.com/rules/axe/4.10/aria-allowed-attr?application=AxeChrome) is used

![image](https://github.com/user-attachments/assets/f1569737-873a-48cf-a213-8921f474fa2a)

We also have a contrast issue with the [dropdown button](https://docs.bokeh.org/en/latest/docs/examples/interaction/widgets/dropdown.html)

For the [date picker](https://docs.bokeh.org/en/latest/docs/examples/interaction/widgets/date_picker.html), the form element isn't labeled.
![image](https://github.com/user-attachments/assets/47df332d-123f-4420-b663-70c8a24926a7)

## Summary
Data visualizations that have some additional forms of interactivity can trigger different sorts of accessibility issues. The axe API found all the issues I have highlighted in this comment and axe-core works with that API. If we decide to keep it, then we can always have checks to make sure that these types of things are not allowed to be introduced into Bokeh. However, if Bokeh doesn't outrightly plan on supporting even more types of interactive components, these ones with issues can be fixed, tested again, and then axe-core can be removed.

I really do hope this gives a bit more context to the poll.

**Reply by bryevdv** [reply]
> However, if Bokeh doesn't outrightly plan on supporting even more types of interactive components, these ones with issues can be fixed, tested again, and then axe-core can be removed.

I think this gets to the crux. It's never safe to assume any fix stays permanent without routine testing to guard against regressions. Based on this, I think it makes sense to keeps these tests in the repo somewhere, and easily runnable.

**Reply by gabalafou** [reply]
Yes, I agree, but I want to sharpen the question here. If there were zero maintenance cost, then adding these tests wouldn't even be in question. But there's never zero maintenance, and the question for me is about whether the net cast by Axe-core is good enough in this case to justify the maintenance burden of adding it.

**Reply by gabalafou** [reply]
I personally lean towards thinking that the maintenance to benefit ratio here is too high, so I would be in favor of documenting the issues found but not integrating Axe-core into the future accessibility testing stack. 

This may not have been made clear, but Axe-core is a conservative set of checks that can be run against web pages to find accessibility violations. It can be thought of loosely like a basic spell checker, but for accessibility. As Chiemezuo has documented, its utility for Bokeh is limited (basically because of `<canvas>`).

How can we help the team make an informed decision re: burden to benefit ratio?

**Reply by bryevdv** [reply]
What kinds of ongoing maintenance is expected if the tools are checked in? Can you give some specific examples?

**Reply by gabalafou** [reply]
Hi @bryevdv, thanks for that question, and thank you so much for engaging on this subject. I'm going to offer some thoughts, but please know that my position is weakly held, and I will defer to your better judgement concerning Bokeh's maintenance. :)

To me, all code is a burden. For example, if somebody were to change the way the docs are built, they might break the Axe-core checks, and then they would have to figure out how to fix them, even though they were working on docs, not accessibility tests. Dependencies also have to be updated every once in a while. 

To be clear, I'm not saying that the maintenance burden is large. I'm emphasizing that the reward is small.

Let's recall two things:

1. Axe-core captures only a fraction of accessibility errors
2. The visual representation of the data that Bokeh provides is opaque to Axe-core because it's rendered with `<canvas>`

This means that Axe-core can only capture a fraction of a fraction of Bokeh accessibility issues.

I also don't get the sense that this fraction of what is being capturing is of great importance when it comes to actual accessibility barriers to using Bokeh. Let's take the slider, for example. Axe-core is complaining because there isn't anything in the markup that unambiguously associates the text above the slider (e.g., Amplitude) with the slider control. In other words, it wants the slider to have an `aria-label` or `aria-labelledby` attribute. While this is undoubtedly a good idea, I don't think it really poses much of an accessibility barrier. This is because a screen reader user navigating the page will likely encounter the title before the slider, so they will know what the slider is for. The more important thing is that the slider is keyboard accessible and communicates its current value to assistive tech, which Bokeh's slider already does (whoever implemented it already was already accessibility-minded, which is awesome!). I'm not saying that it's not important to label the slider. One reason that comes immediately to mind is that screen reader software provides a way to bring up all of the form controls on the page, but if they are not labelled in an accessible way, then this form control quick nav is pretty useless. And I'm sure there are other reasons. But I don't believe it's a showstopper, more like a polish task. 

We can take another example, the lack of an accessible name for the Bokeh link icon in the toolbar. In my mind, it's really a minuscule difference that the link reads out "https://bokeh.org" instead of "Bokeh homepage" or something. 

And then some things are understated. Like the date picker. The real problem with the date picker is not that it doesn't have an accessible name; it's that it's impossible to use with just a keyboard, and it's challenging to use with a screen reader.

In other words, I feel like this tool, Axe-core, is a bit of a distraction. I think it was worthwhile to investigate its potential. Along the way we discovered and documented its limitations, all of which is valuable. 

What I am not sure of is that it is valuable to add it to the codebase. There is the maintenance question, yes, and there's also the fact that this is not quite ready to ship, so there's still some amount of work to do before this is ready for merge. For example, we need to make it clear that the approach depends on the docs being built without actually doing the docs-build step in the test. 

If more of Bokeh was testable by Axe-core, I would be in favor of adding it. Also I want to be clear that this is not about Playwright. I still want to add Playwright to the repo. I'm just not sure about this idea of build the docs and pass each example from the docs through Axe-core.

**Reply by gabalafou** [reply]
Oh, I should add, for balance and fairness, that the I do think the contrast issues discovered are worthwhile. Perhaps for that reason alone, it might be worth adding.

**Reply by frankelavsky** [reply]
I'd love to jump in to the discussion with my opinion on this, as well as some suggestions for paths forward. (For context: I was asked to give this conversation and #14032 a read-through, which is why this is a bit on the longer side. I tried to organize all my thoughts below as much as possible, while still being thorough.)

## Short summary
I largely agree with @gabalafou: the benefit of axe-core is minor here, since the major issues that matter are items we identified in our audit and are not currently within the scope of what automated tooling can capture reliably. I think we should document the findings here that are worth taking action on (if we haven't already), check to see if any of the issues addressed are already part of our roadmap and path forward, and (if not) consider whether to include those issues in our roadmap. Lastly, I do not recommend including axe-core right now. 

## Pros/cons (in my opinion)

### Pros of keeping axe-core in test setup

- **Regression prevention for widget accessibility** - Once issues like unlabeled sliders and contrast problems are fixed, automated tests prevent them from being reintroduced... that means that if changes are made that happen to fall within the scope of the tests, we can hopefully avoid regression. (How common is it to do contrast/semantics changes on existing UI elements? This pro's value increases the more the UI is actually changed and updated in Bokeh.)
- **Catches expanding scope** - If Bokeh adds more HTML-based widgets/controls in the future, axe-core coverage automatically expands (same caveat as before is here: how common is it to add new HTML UI to the ecosystem?)
- **Low barrier for contributors** - New contributors might not know accessibility best practices; automated checks could catch fundamental issues before merging. Again, the value of this is dependent on how common it is for new community members to contribute to UI-related changes (or add new HTML elements).
- **"Documents" expectations** - Having tests codifies what "accessible" means for Bokeh's non-canvas elements, serving as living documentation. I love testing for this reason, because it is a sort of documentation/rules that are discovered via usage and as-needed.
- **Keyboard nav/AT automation is awesome** - Playwright's major strength in testing environments is related to automated testing for tab navigation and catching keyboard traps. (see con for this below, though)

### Cons of keeping axe-core in test setup

- **Keyboard nav/AT automation might be complex** - Because bokeh is going to have some changes to keyboard access, I am confident that we should not implement our tests for this until we lock in on the expected, low-level behaviors for things. But also, establishing good testing for an environment like Bokeh might end up being very complex (navigable chart elements, for example, will be hard to build rich tests for if Data Navigator isn't going to be a core part of Bokeh, but just a plugin). I haven't done enough with playwright to know how much work this would _actually_ entail though, so please keep this particular con in mind with a grain of salt.
- **False sense of security** - Having accessibility tests might make the team think accessibility is "handled" when canvas elements (the core product) are completely untested. We don't want our devs (or dev-users) to get a false sense of hope that Bokeh is locked in and accessible to some level of standard. (However, clear docs could just mitigate this, so it isn't a big issue per se.)
- **Mismatch between tool and problem domain** - For better or worse, axe-core is designed for document-based websites, not canvas-heavy visualization libraries. I am not sure that canvas-related heuristics or models are even on the roadmap for the axe-core devs at all, let alone ones specific to visualization.
- **Test fragility** - I am really concerned with @mattpap's comment about dependence on the docs; coupling accessibility tests to docs build process creates unnecessary interdependency... ideally there are isolated (non-docs) environments for testing. Changes to docs infrastructure, example formatting, or build tooling could break these tests, requiring developers to fix accessibility tests when doing unrelated work. And building out isolated test environments is its own maintenance problem (and I'd argue much too great of a cost for the benefit). I understand why piggybacking on the docs was done (this is common), but if the docs aren't going to be reliable in the future, the tests may end up fragile, outdated, and full of false positives. (And please correct me if I misunderstand this?)
- **Opportunity cost** - Time spent maintaining axe-core integration could be spent on manual testing of keyboard navigation, screen reader compatibility, or alternative text representations of data (which is much higher-impact accessibility work). Audits seem costly, but that's only because we have n=1 right now. Future audits will (hopefully) be a much lower lift and still provide meaningful benefits for the cost.
- **No clear ownership** - In long-running projects, who maintains this? If it's not core maintainers' priority and not blocking real users, I do worry that this becomes technical debt or will magnify it down the road (see later, related con on "maintenance without dedicated folks")
- **Doesn't test what matters most** - For example (as mentioned by @gabalafou), the date picker being keyboard-inaccessible is a real barrier; it not having an aria-label is relatively trivial. Axe is a deterministic tool with limited scope. And for better or worse, the complexity of access barriers that can arise from an interactive visualization library are just not well-serviced by tools like axe right now.
- **Limited upstream ecosystem** - Other canvas-based visualization libraries don't use axe-core (that I am aware of), so Bokeh would be the first (this could be a good thing? maybe start a relationship with deque? but then that would be more work for someone). But we don't have examples to learn from, since axe can't even handle canvas visualizations right now. Down the road, this might change, and some libraries might start using it. In that case, following in their footsteps will be a lower lift of effort.
- **Creates arbitrary/unequal standards for contribution** - New features with widgets must pass axe-core even though core visualizations can't, creating inconsistent standards. Visualizations should actually have the most scrutiny, if anything. (That being said, this might not be a proper con for using axe but just more of an acknowledgement of what we really could use)
- **Maintenance without dedicated accessibility folks** - If Bokeh doesn't have dedicated accessibility resources, automated tests become "someone else's problem" until they break critically. This, to me, is a big issue! Either we need to have really well-documented information for passing the torch to new contributors (perhaps adding to the contribution governance?) or have some dedicated folks in place. If not, I have a feeling the moment this causes a barrier for someone down the road, it will be an issue that needs to be hashed out by someone (and probably eventually find its way to this discussion here)

## Longer-form thoughts on axe
While automatically catching contrast and aria/semantics is great (and having some kind of CI hook that at least surfaces warnings is useful), I'd argue that we have largely caught most of these in our audit. This means that moving forward, the value would have to be interpreted and understood by the devs, including all of the caveats that can come with axe in an environment like Bokeh. Devs will need to keep in mind what limitations testing has and (perhaps _more_ importantly) will need to know that "passing" tests is hardly a state worth celebrating. The value of axe in Bokeh contributes, what I believe, already-well-considered issues (contrast/semantics), meaning that we might end up building a sense of false confidence in our devs if they see accessibility "passing" on the relatively easy and well known issues.

Environments that gain significant value from axe-core are ones where _new_ interfaces are being built using existing or new components. So, in the case where a new UI widget gets integrated, having a testing suite in place would be valuable. However, the state of Bokeh is such that I think adding new widgets are far less common (correct me if I am wrong @mattpap). The real value of axe-core would be for the users of bokeh, who might change CSS/styling or other cosmetics once they end up embedding Bokeh into whatever they are building.

**Reply by frankelavsky** [reply]
And just to be clear: my major gripe is about whether the value is worth it (both now and in the future, once maintenance is required). And in an ideal situation, we are auditing Bokeh to catch both simple and complex issues until (or _if_) automated testing improves, which feels like the benefit here is largely going to be redundant if auditing is something that is still done going forward.
